codice/ddf
# DDF-2120: Provide sanitized reasons for failures on the query that is in progress

#### What does this PR do?
(1) <s>Cleans up some ReliableResource code without impacting threading. That will be done in a second ticket</s>. This component is not immediately relevant to the ticket and has been removed. 
(2) Adds a list of reasons (String's) to the QueryStatus object for population during a failed query. 
(3) <s>The Search object recursively traverses a stack trace and loads the reasons</s>. The search object generates a safe error message based on the exception type or causality types for each Processing Detail that reports failure. 

<s>Right now we do a comprehensive collection of messages from the stack trace. We only grab the top level message on the stack trace</s>. We use the exception type to generate a sanitized error message that does not expose implementation details. 

#### Who is reviewing it?
@figliold 
@shaundmorris 

#### Choose 2 committers to review/merge the PR:
@lessarderic 
@pklinef

#### How should this be tested?
(1) Full build for DDF + other layers to ensure nothing broke (PRBS will do)
(2) Afterwards, spin up a clean DDF and start a copy of the `cometd-web-example` by running the following in the example's root directory: `npm i && grunt`. Open a browser to `http://localhost:8000/`.
(3) In the cometD client, with DDF fully setup, run a null query. Assert that in the response JSON (the last response received; which is the third one from the top if no extra sources were configured) matches the below snippet, particularly the populated `reasons` field (figure A). Run a valid query and assert that the `reasons` field is now empty (figure B). 

###### Figure A
```
{
    "data": {
        "hits": 0,
        "metacard-types": {},
        "id": "0a8de366-0657-c3bf-ab70-00fde17ec1cd",
        "results": [],
        "status": [
           {
                "hits": 0,
                "elapsed": 211,
                "reasons": [
                    "Query was invalid"
                ],
                "id": "ddf.distribution",
                "state": "FAILED",
                "results": 0
            }
        ],
        "successful": true
    },
    "channel": "/0a8de366-0657-c3bf-ab70-00fde17ec1cd"
}
```

###### Figure B
```
{
    "data": {
        "hits": 0,
        "metacard-types": {},
        "id": "9d1d796c-83a4-5e6a-d3db-55061c0479f3",
        "results": [],
        "status": [
           {
                "hits": 0,
                "elapsed": 133,
                "reasons": [],
                "id": "ddf.distribution",
                "state": "SUCCEEDED",
                "results": 0
            }
        ],
        "successful": true
    },
    "channel": "/9d1d796c-83a4-5e6a-d3db-55061c0479f3"
}
```
#### Any background context you want to provide?
Need feedback on the over-all strategy, plus any components that might be missing or not considered, before the tests are written. 

Relevant implementation area in the [source code](https://github.com/codice/ddf/blob/master/catalog/ui/search-ui/search-endpoint/src/main/java/org/codice/ddf/ui/searchui/query/model/Search.java#L144-L175). 

#### What are the relevant tickets?
DDF-2120

#### Checklist:
